raft/src/main/java/org/apache/kafka/raft/RaftClient.java

```diff
@@ -80,8 +81,11 @@ interface Listener<T> {
          * epoch.
          *
          * @param leader the current leader and epoch
+         * @param endOffset the current log end offset (exclusive). This is useful for nodes that
+         *                  are claiming leadership, because it lets them know what log offset they
+         *                  should attempt to write to next.
```

Comment:
This is only possible in the leader. This is not well defined on any replica outside of the leader. What problem are you trying so solve?

KRaft guarantees that the committed offset is the LEO when notifying the leader that is now the leader of the partition. For all other replicas leadership change and lost of leadership is done immediately.

Reply:
> What problem are you trying so solve?

We have to know the offset of records that we apply. But we apply records before we submit them to Raft.